SI-6196 - Set should implement filter

rklaehn opened the pull request:
Implements a version of filter and filterNot that reuses as much as
possible from the existing tree instead of building an entirely new one
like the builder-based filter does. This results in significant performance
improvements on average.

Adds a test of basic correctness of filter and filterNot as well as of the
desirable properties of the new filter implementation.

This is a collaboration between me and @Ichoran

Review by @Ichoran, @soc

rklaehn commented:
Here is an example that creates an OOME with the current implementation of filter and will work just fine with the new one due to the improved structural sharing:

val n = 10000
val s = (0 until n).toSet
val r = (0 until n).scanLeft(s) { case (s,i) => s.filter(_ > i) }

Performance is increased about a factor of 2.5 over the old approach even for a test case that is designed to give the old builder-based approach an advantage.

rklaehn commented:
Here is a simple benchmark. For this I get a factor of 4 improvement!

--- code ---
  def linearBenchmark(n:Int) {

```
def work(s:Set[Int]) : Int = {
  var result = 0
  for(i<-0 until s.size by s.size / 100) {
    result += s.filter(_<i).size
  }
  result
}

def tToS(time:Double) = (time*1000000).formatted("%.3f")

def rToS(r:Double) = r.formatted("%.3f")

val a = HashSet.empty[Int] ++ (0 until n)
val b = HashSet2.empty[Int] ++ (0 until n)
val (r0, t0) = th.benchPairWarm(() => work(a))
val (r1, t1) = th.benchPairWarm(() => work(b))

val time0 = tToS(t0.runtime)
val time1 = tToS(t1.runtime)
val ratio = rToS(t0.runtime/t1.runtime)
println(s"$time0\t$time1\t$ratio")
```

  }
  linearBenchmark(10000)

--- result ---

73459.037   18033.048   4.074

Ichoran commented:
I was using the wrong format for pull requests; the title should be (if fixing a bug):

  SI-#### Title of bug from bug report

FYI.  I haven't reviewed the changes yet, but this looks easier to maintain than my version.

Ichoran commented:
Aside from the title, LGTM. On my machine this isn't consistently faster than my implementation, but I think it is enough easier to understand that it's not worth it to try too hard to decide which of the two is faster.

rklaehn commented:
I modified the title of the PR. Do I also need to modify the title of the commit?

If I do a similar PR for HashMap, there will be some code duplication (the keepBits method). I guess that is unavoidable for now if we want two independent PRs. The keepBits method could then be factored out into an utility class (TrieUtils?) in the next step

rklaehn commented:
One more thing: with filter and filterNot now being more efficient, other set operations can be made to take advantage of it. intersect is already implemented using filter, so it automatically profits from this. 

Diff currently is implemented in Subtractable and removes the elements one by one, so it is not very efficient. This is easy to fix:

  override def diff(that:GenSet[A]) = this filterNot that

Should this be part of this PR, or a separate one? Also, should -- use diff in case that is a HashSet?

retronym commented:
Please change the commit title; that's the most important place to follow the conventions in the contributors guide. Drop the ':' too.

Sounds like the followon changes are closely related and would benefit from a holistic review, so feel free to include them as additional commits on this PR.

rklaehn commented:
I added taking advantage of the new filter/filterNot in HashSet.diff and HashSet.-- in separate commits as requested. I also added the implementation of the same algorithm in HashSet (SI-6200) as a separate commit.

Ichoran commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -64,6 +66,18 @@ class HashMap[A, +B] extends AbstractMap[A, B]
+    nullToEmpty(filter0(p, true, 0, buffer, 0))
+  }
+
+  protected def filter0(p: ((A, B)) => Boolean, negate:Boolean, level: Int, buffer: Array[HashMap[A, B @uV]], offset0: Int): HashMap[A, B] = null
```
Space after colon missing

Ichoran commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -203,6 +217,9 @@ object HashMap extends ImmutableMapFactory[HashMap] with BitOperations.Int {
     override def removed0(key: A, hash: Int, level: Int): HashMap[A, B] =
       if (hash == this.hash && key == this.key) HashMap.empty[A,B] else this
 
+    override protected def filter0(p: ((A, B)) => Boolean, negate:Boolean, level: Int, buffer: Array[HashMap[A, B @uV]], offset0: Int): HashMap[A, B] =
```
Space after colon missing (negate)

Ichoran commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -342,6 +374,52 @@ object HashMap extends ImmutableMapFactory[HashMap] with BitOperations.Int {
       }
     }
 
+    override protected def filter0(p: ((A, B)) => Boolean, negate:Boolean, level: Int, buffer: Array[HashMap[A, B @uV]], offset0: Int): HashMap[A, B] = {
```
Space after colon (negate)

Ichoran commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -445,6 +523,50 @@ object HashMap extends ImmutableMapFactory[HashMap] with BitOperations.Int {
+   * @param size the maximum size of the collection to be generated
+   * @return the maximum buffer size
+   */
+  @inline private def bufferSize(size:Int) : Int = (size + 6) min (32 * 7)
```
Space after colon (size)

Ichoran commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -279,6 +310,52 @@ object HashSet extends ImmutableSetFactory[HashSet] {
+      var i = 0
+      while (i < elems.length) {
+        val result = elems(i).filter0(p, negate, level + 5, buffer, offset)
+        if (result ne null) {
```
Not sure the comments here really help anything.  What could buffer(offset) = result do except add the result to the buffer?  Better to just remove the comments and have one statement per line (i.e. put offset += 1 on its own line)

Ichoran commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -292,6 +369,50 @@ object HashSet extends ImmutableSetFactory[HashSet] {
+   * @param size the maximum size of the collection to be generated
+   * @return the maximum buffer size
+   */
+  @inline private def bufferSize(size:Int) : Int = (size + 6) min (32 * 7)
```
Colon, space (one too few, one too many)

Ichoran commented:
Looks basically fine save for cosmetic stuff.  Will merge once that's done.

rklaehn commented:
Sorry about the inconsistent indentation. I was in a hurry to make a PR before the deadline.

rklaehn commented:
corrected missing newline at the end of t6196.scala

Ichoran commented:
Can you rebase etc. so this merges cleanly?

rklaehn committed:
SI-6196 - Set should implement filter

Implements a version of filter and filterNot that reuses as much as
possible from the existing tree instead of building an entirely new one
like the builder-based filter does. This results in significant performance
improvements on average.

Adds a test of basic correctness of filter and filterNot as well as of the
desirable properties of the new filter implementation.

This is a collaboration between me and @Ichoran

rklaehn committed:
SI-6200 - HashMap should implement filter

This is the exact same algorithm as in SI-6196, with only slight
differences due to the two type arguments of HashMap.

Filter is tested by the new comparative collection test by @Ichoran, but
there is nevertheless a small correctness test in t6200 in addition to
tests of the new desirable behavior.
